cmd/soroban-cli/src/events.rs

```diff
@@ -0,0 +1,437 @@
+            std::usize::MAX
+        } else {
+            self.count
+        };
+
+        // Read the JSON events from disk and find the ones that match the
+        // contract ID filter(s) that were passed in.
+        Ok(snapshot::read_events(path)
+            .map_err(|err| Error::CannotReadFile {
+                path: path.to_str().unwrap().to_string(),
+                error: err.to_string(),
+            })?
+            .iter()
+            // FIXME: We assume here that events are read off-disk in
+            // chronological order, but we should probably be sorting by ledger
+            // number (and ID, for events within the same ledger), instead,
+            // though it's likely that this logic belongs more in
+            // `snapshot::read_events()`.
+            .rev()
+            // The ledger range is optional in sandbox mode.
+            .filter(|evt| {
+                if self.start_ledger == 0 && self.end_ledger == 0 {
+                    return true;
+                }
+
+                match evt.ledger.parse::<u32>() {
+                    Ok(seq) => seq >= self.start_ledger && seq <= self.end_ledger,
+                    Err(e) => {
+                        eprintln!("error parsing key 'ledger': {:?}", e);
+                        eprintln!(
+                            "your sandbox events file ('{}') may be corrupt",
+                            path.to_str().unwrap(),
+                        );
+                        eprintln!("ignoring this event: {:#?}", evt);
+
+                        false
+                    }
+                }
+            })
+            .filter(|evt| {
+                // Contract ID filter(s) are optional, so we should render all
+                // events if they're omitted.
+                self.contract_ids.is_empty()
+                    || self.contract_ids.iter().any(|id| *id == evt.contract_id)
+            })
+            .filter(|evt| {
+                // Like before, no topic filters means pass everything through.
+                self.topic_filters.is_empty() ||
+                // Reminder: All of the topic filters are part of a single
+                // filter object, and each one contains segments, so we need to
+                // apply all of them to the given event.
+                self.topic_filters
+                    .iter()
+                    // quadratic but both are <= 5 long
+                    .any(|f| does_topic_match(&evt.topic, &f.split(',').collect()))
+            })
+            .take(count)
+            .cloned()
+            .collect::<Vec<Event>>())
+    }
+}
+
+// Determines whether or not a particular filter matches a topic based on the
+// same semantics as the RPC server:
+//
+//  - for an exact segment match, the filter is a base64-encoded ScVal
+//  - for a wildcard, single-segment match, the string "*" matches exactly one
+//    segment
+//
+// The expectation is that a `topic_filter` is a comma-separated list of
```

Comment:
*style*
do you mean?
```suggestion
// The expectation is that a `segments` is a comma-separated list of
```